packages/headless/src/utils/expression-builder-sapi-v4.ts

```diff
@@ -0,0 +1,187 @@
+type QueryExpresion =
+  | KeyWord
+  | ExactMatch
+  | FieldExpresion
+  | Near
+  | Not
+  | FieldExists
+  | ObjectAccess
+  | QueryExtensionInvocation
+  | QuerySyntax
+  | And
+  | Or;
+
+type FieldValue = DateRange | Date | KeyWord | ExactMatch;
+
+type TextValue = KeyWord | ExactMatch;
+
+interface DateRange {
+  type: 'dateRange';
+  dateRange: {
+    from: string;
+    to: string;
+  };
+}
+
+interface Date {
+  type: 'date';
+  dateValue: string;
+}
+
+interface KeyWord {
+  type: 'keyword';
+  value: string;
+}
+
+interface ExactMatch {
+  type: 'exactMatch';
+  value: string;
+}
+
+interface And {
+  type: 'and';
+  expressions: QueryExpresion[];
+}
+
+interface Or {
+  type: 'or';
+  expressions: QueryExpresion[];
+}
+
+interface FieldExpresion {
+  type: 'fieldExpression';
+  fieldName: string;
+  operator:
+    | 'contains'
+    | 'isExactly'
+    | 'lowerThan'
+    | 'lowerThanOrEqual'
+    | 'greaterThan'
+    | 'greaterThanOrEqual'
+    | 'phoneticMatch'
+    | 'fuzzyMatch'
+    | 'regexMatch'
+    | 'wildcardMatch'
+    | 'differentThan';
+  value: FieldValue; // shouldn't this be an array?
+}
+
+interface Near {
+  type: 'near';
+  startTerm: TextValue;
+  otherTerms: OtherTerm[];
+}
+
+interface OtherTerm {
+  maxKeywordsFrom: number;
+  endTerm: TextValue;
+}
+
+interface Not {
+  type: 'not';
+  expression: QueryExpresion;
+}
+
+interface FieldExists {
+  type: 'fieldExists';
+  fieldName: string;
+}
+
+// Not sure how to use this one.
+interface MatchAll {
+  type: 'matchAll';
+}
```

Comment:
@acote-coveo, how should this object be used? It only has a `type` property in the doc, hmm.

Reply:
It also have a value of type string.